src/input/input_python.rs

```diff
@@ -307,47 +307,37 @@ impl<'py> Input<'py> for Bound<'py, PyAny> {
         Err(ValError::new(ErrorTypeDefaults::FloatType, self))
     }
 
-    fn strict_decimal(&self, py: Python<'py>) -> ValResult<Bound<'py, PyAny>> {
+    fn validate_decimal(&self, strict: bool, py: Python<'py>) -> ValMatch<Bound<'py, PyAny>> {
         let decimal_type = get_decimal_type(py);
-        // Fast path for existing decimal objects
-        if self.is_exact_instance(decimal_type) {
-            return Ok(self.to_owned());
-        }
-
-        // Try subclasses of decimals, they will be upcast to Decimal
-        if self.is_instance(decimal_type)? {
-            return create_decimal(self, self);
-        }
-
-        Err(ValError::new(
-            ErrorType::IsInstanceOf {
-                class: decimal_type
-                    .qualname()
-                    .and_then(|name| name.extract())
-                    .unwrap_or_else(|_| "Decimal".to_owned()),
-                context: None,
-            },
-            self,
-        ))
-    }
 
-    fn lax_decimal(&self, py: Python<'py>) -> ValResult<Bound<'py, PyAny>> {
-        let decimal_type = get_decimal_type(py);
         // Fast path for existing decimal objects
         if self.is_exact_instance(decimal_type) {
-            return Ok(self.to_owned().clone());
-        }
-
-        if self.is_instance_of::<PyString>() || (self.is_instance_of::<PyInt>() && !self.is_instance_of::<PyBool>()) {
-            // checking isinstance for str / int / bool is fast compared to decimal / float
-            create_decimal(self, self)
+            Ok(ValidationMatch::exact(self.to_owned().clone()))
         } else if self.is_instance(decimal_type)? {
-            // upcast subclasses to decimal
-            return create_decimal(self, self);
-        } else if self.is_instance_of::<PyFloat>() {
-            create_decimal(self.str()?.as_any(), self)
+            // Upcast subclasses to decimal
+            create_decimal(self, self).map(ValidationMatch::strict)
```

Comment:
This was deliberately below the `PyString` / `PyInt` / `PyBool` check because those are faster to check for (see comment now on line 334), maybe move those up (but have them with a `!strict` guard too).